### PR TITLE
Public sequence, long ConsumedBytes

### PR DESCRIPTION
**Pull request by pakrym**
Public `Sequence` allows some nice extension methods on `BufferReader` for example `SliceToEnd()`

https://github.com/dotnet/corefxlab/issues/2019

### CODE REVIEW

**Comment by pakrym** [comment]
Closing in favour of larger https://github.com/dotnet/corefxlab/pull/2024

**Comment by pakrym** [comment]
That PR got closed, reopening.

**Comment by KrzysztofCwalina** [comment]
I don't think we should be exposing sequence (or any other APIs) without validating that they are needed and sufficient to implement the extensions methods.

**Comment by pakrym** [comment]
Reader doesn't have anything else useful that we don't expose. Having sequence is enough for what I think of: I want to implement a method that returns `ROB` from the current position of `Reader` to the end.

What's the reason not to expose it?

**Comment by KrzysztofCwalina** [comment]
The main reason to not expose it is that we need to validate. You are adding APIs that we don't exercise. Moreover, I suspect raw sequence exposed will be very error prone as extensions will have to remember how much they already read.

Secondary reason is that I just implemented many extension methods and did not have to use the sequence.

**Comment by pakrym** [comment]
The extension would be for `BufferReader<ReadOnlyBuffer>` not `BufferReader<T>` and use Position to Slice to the end.

I've looked at some example of how BuffferReader is used and there seems to be a popular theme of wanting to slice underlying buffer using current position after some reading.

**Comment by pakrym** [comment]
> Secondary reason is that I just implemented many extension methods and did not have to use the sequence.
So how would you implement `Slice to the end of underlying buffer` without having access to buffer itself?

**Comment by KrzysztofCwalina** [comment]
We need to add the following primitives (as instance methods) to the reader:
PositionOf(byte)
PositionOf(ReadOnlySpan<byte>)
AdvanceTo(Position)

**Comment by pakrym** [comment]
How is it related?

**Comment by KrzysztofCwalina** [comment]
> How is it related?

Sorry, I was answering: "So how would you implement Slice to the end of underlying buffer without having access to buffer itself?"

**Comment by pakrym** [comment]
I still don't see how to do it even having those methods.

My implementation would look like this:

```
public static ReadOnlyBuffer SliceToEnd(ref this BufferReader<ReadOnlyBuffer> reader)
{
    return reader.Sequence.Slice(reader.BytesConsumed);
}
```

**Comment by KrzysztofCwalina** [comment]
Your implementation would not compile. Sequence does not have a method Slice. Which kind of proves my point that we need to add the property together with actual extensions that we hope to implement.

**Comment by pakrym** [comment]
It is on `BufferReader<ReadOnlyBuffer>`.

**Comment by KrzysztofCwalina** [comment]
> It is on BufferReader<ReadOnlyBuffer>

This is an indication that the BufferReader<T> abstraction is not complete.

**Comment by pakrym** [comment]
No, it's an indication that ISequence abstraction is not complete. And it's impossible to build everything on top of it, you at lease need Slice.

**Comment by KrzysztofCwalina** [comment]
Fair, either of the abstractions is not complete. Which again proves my point that we need to add the extension methods that we hope to implement in the same PR as the APIs we are adding to implement such extensions.

**Comment by pakrym** [comment]
Added extension methods and tests.

File: src/System.Text.Http.Parser/HttpParser.cs [hunk 1]
```diff
@@ -356,7 +356,7 @@ public HttpParser(bool showErrorDetails)
             finally
             {
                 consumed = reader.Position;
-                consumedBytes = reader.ConsumedBytes;
+                consumedBytes = (int)reader.ConsumedBytes;
```
Comment (mgravell) [inline src/System.Text.Http.Parser/HttpParser.cs]: consider `checked((int)reader.ConsumedBytes);` ?

File: tests/Benchmarks/BytesReaderBench.cs [hunk 2]
```diff
@@ -84,7 +84,7 @@ static void ParseInt32ReadableBufferReader()
 
             using (iteration.StartMeasurement())
             {
-                while(Utf8Parser.TryParse(reader.Span.Slice(reader.ConsumedBytes), out int value, out int consumed)){
+                while(Utf8Parser.TryParse(reader.Span.Slice((int)reader.ConsumedBytes), out int value, out int consumed)){
```
Comment (mgravell) [inline tests/Benchmarks/BytesReaderBench.cs]: consider checked cast

**Comment by pakrym** [comment]
Closing, pending design decisions,

File: tests/System.Buffers.Primitives.Tests/ReadableBufferReaderFacts.cs [hunk 3]
```diff
@@ -84,6 +84,34 @@ public void CursorIsCorrectAtEnd()
             reader.Take();
             Assert.True(reader.End);
         }
+        
+        [Fact]
+        public void SliceAftereWorks()
+        {
```
Comment (bruno-garcia) [inline tests/System.Buffers.Primitives.Tests/ReadableBufferReaderFacts.cs]: typo: SliceAfter**e**Works -> SliceAfterWorks

**Pull request closed without merging**